src/main/java/com/chatapp/vartalaab/service/MessageService.java

```diff
@@ -0,0 +1,73 @@
+package com.chatapp.vartalaab.service;
+
+import com.chatapp.vartalaab.dto.MessageDto;
+import com.chatapp.vartalaab.redisEntity.MessageEntity;
+import com.chatapp.vartalaab.redisEntity.OfflineMessageIds;
+import com.chatapp.vartalaab.repository.MessageRepository;
+import com.chatapp.vartalaab.repository.OfflineMessageIdsRepository;
+import com.chatapp.vartalaab.utils.GeneralUtility;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.web.socket.TextMessage;
+import org.springframework.web.socket.WebSocketSession;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Service
+@Slf4j
+public class MessageService {
+
+    private MessageRepository messageRepository;
+
+    private OfflineMessageIdsRepository offlineMessageIdsRepository;
+
+    private UserSessionService userSessionService;
+
+    @Autowired
+    private ConcurrentHashMap<String, WebSocketSession> sessionMap;
+
+    public MessageService(MessageRepository messageRepository, OfflineMessageIdsRepository offlineMessageIdsRepository, UserSessionService userSessionService){
+        this.messageRepository = messageRepository;
+        this.offlineMessageIdsRepository = offlineMessageIdsRepository;
+        this.userSessionService = userSessionService;
+    }
+
+    public List<String> getMessagesFromCache(String username) throws IOException {
+        log.debug(">> getMessagesFromCache");
+        List<String> messages = new ArrayList<String>();
+        Optional<OfflineMessageIds> offlineMessageIds = offlineMessageIdsRepository.findById(username);
+        if(offlineMessageIds.isPresent()) {
+            List<String> messageIds = offlineMessageIds.get().getMessageIds();
+            for (String messageId : messageIds) {
+                MessageEntity message = messageRepository.findById(messageId).orElseThrow(() -> new NoSuchElementException());
+                MessageDto messageDto = new MessageDto(message.getSender(), message.getMessage(), message.getTimestamp());
+                messages.add(messageDto.toString()); //returning JSON message in text format
+                this.forwardTheMessageToReceiver(messageDto.getSender(), new TextMessage(GeneralUtility.AckForReceived));
+                messageRepository.deleteById(messageId);
```

Comment:
If my understanding is correct, the msgid should also get deleted from OfflineMessagesIds right ?

Reply:
yes, added that in code now.